File: assets/hw/template/autograder/hw_release_sln/python/questions/ch10_ex9.py
import collections
import os
from typing import Any, Dict, List, Text, Tuple

from absl import logging
import matplotlib.pyplot as plt
import pandas as pd
from scipy import cluster
from sklearn import preprocessing


class USArrests(object):
  """Chapter 10, Exercise 9."""

  def __init__(self, data_path: Text="../data/USArrests.csv",
               figsize: Tuple[int, int]=(10, 10)):
    """Constructor"""
    self.data_path = data_path
    assert os.path.exists(self.data_path), "USArrests dataset not found."
    logging.info("\tLoading %s", self.data_path)
    self.us_arrests = pd.read_csv(self.data_path, index_col='Unnamed: 0')
    self.figsize = figsize

  def part_a(self) -> Dict[Text, Any]:
    """
    Using the cluster.hierarchy.linkage() and cluster.hierarchy.dendrogram()
    functions, return a figure of the states clustered using hierarchical
    clustering with complete linkage and Euclidean distance.

    n.b. To reuse the cluster results in subsequent questions, please save
    the cluster fit as an attribute, i.e. self.clusters = ...

    Returns:
      A dict of the dendogram values required for the plot.
    """
    self.clusters = cluster.hierarchy.linkage(self.us_arrests, method='complete')
    _, ax = plt.subplots(figsize=self.figsize)
    dendo = cluster.hierarchy.dendrogram(self.clusters,
                                         orientation='top',
                                         labels=self.us_arrests.index,
                                         distance_sort='descending',
                                         show_leaf_counts=True,
                                         ax=ax)
    return dendo

  def part_b(self) -> Dict[int, List[Text]]:
    """
    Using the cluster.hierarchy.cut_tree() function, cut the previously fit
    dendogram into three distinct clusters.

    Returns:
      A dict containing lists of states for each cluster, keyed to integers
      representing the cluster number.
    """
    if not hasattr(self, 'clusters'):
      raise ValueError("clusters not found")
    clusts = cluster.hierarchy.cut_tree(self.clusters, n_clusters=[3])
    groups = collections.defaultdict(list)
    for state, group in zip(self.us_arrests.index, clusts[:,0]):
      groups[group].append(state)

    return groups

  def part_c(self):
    """
    Using the cluster.hierarchy.linkage() and cluster.hierarchy.dendrogram()
    functions, return a figure of the states clustered using hierarchical
    clustering with complete linkage and Euclidean distance after after
    scaling the variables.

    n.b. You can use the preprocessing function from sklearn to do the scaling.

    Returns:
      A dict of the dendogram values required for the plot.
    """
    us_arrests_std = pd.DataFrame(preprocessing.scale(self.us_arrests, axis=0),
                                  columns=self.us_arrests.columns)
    clusters = cluster.hierarchy.linkage(us_arrests_std, method='complete')
    _, ax = plt.subplots(figsize=self.figsize)
    dendo = cluster.hierarchy.dendrogram(clusters,
                                         orientation='top',
                                         labels=self.us_arrests.index,
                                         distance_sort='descending',
                                         show_leaf_counts=True,
                                         ax=ax)
    return dendo


def ch10_ex9():
  logging.info("Chapter 10, Exercise 9.")
  us_arrests = USArrests()

  logging.info("\tPart a.")
  us_arrests.part_a()
  plt.savefig("../results/python/ch10_ex9_part_a.png")

  logging.info("\tPart b.")
  groups = us_arrests.part_b()
  with open("../results/python/ch10_ex9_part_b.txt", "w") as f:
    for group, states in groups.items():
      f.write("Group " + str(group) + ":\n")
      f.write(",".join(states))
      f.write("\n")

  logging.info("\tPart c.")
  us_arrests.part_c()
  plt.savefig("../results/python/ch10_ex9_part_c.png")

File: assets/hw/template/autograder/hw_release_sln/R/questions/ch10_ex9.R
# Chapter 10, Exercise 9.

suppressPackageStartupMessages({
  library(logging)
})

#' The width and height of exported plots.
FIGSIZE = c(1000, 1000)


#' Loads required data for exercise.
#'
#' @param data_path location of the data.
#' @return a data.frame containing the exercise data.
LoadData <- function(data_path="../data/USArrests.csv", verbose=TRUE) {
  stopifnot(file.exists(data_path))
  if (verbose)
    loginfo(sprintf("\tLoading %s", data_path))
  us_arrests <- read.csv(data_path)
  rownames(us_arrests) <- us_arrests$X
  us_arrests$X <- NULL
  return (us_arrests)
}


#' Using hierarchical clustering with complete linkage and
#' Euclidean distance, cluster the states.
#'
#' @param data_frame a data.frame containing the US arrests data.
#' @return The dendogram plot.
PartA <- function(data_frame) {
  # YOUR CODE STARTS HERE
  d <- dist(data_frame, method="euclidean")
  h <- hclust(d, method="complete")
  fig <- plot(h)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (fig)
}


#' Cut the dendrogram at a height that results in three distinct
#' clusters. Which states belong to which clusters?
#'
#' n.b. You'll need to refit the cluster from Part A to complete this part.
#'
#' @param data_frame a data.frame containing the US arrests data.
#' @return A list containing a vector of state names for each cluster.
PartB <- function(data_frame) {
  # YOUR CODE STARTS HERE
  d <- dist(data_frame, method="euclidean")
  h <- hclust(d, method="complete")
  tree_clusts <- cutree(h, 3)
  clusters <- list()
  for(i in 1:3) {
    clusters[[i]] = names(tree_clusts[tree_clusts==i])
  }
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (clusters)
}


#' Hierarchically cluster the states using complete linkage and Euclidean
#' distance, after scaling the variables to have standard de- viation one.
#'
#' @param data_frame a data.frame containing the US arrests data.
#' @return The dendogram plot.
PartC <- function(data_frame) {
  # YOUR CODE STARTS HERE
  d <- dist(scale(data_frame), method="euclidean")
  h <- hclust(d, method="complete")
  fig <- plot(h)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (fig)
}


#' Wrapper function for Chapter 3, Exercise 9.
Ch10Ex9 <- function() {
  loginfo("Chapter 10, Exercise 9.")
  us_arrests = LoadData()

  loginfo("\tPart a.")
  png("../results/R/ch10_ex9_part_a.png", width=FIGSIZE[1], height=FIGSIZE[2])
  PartA(us_arrests)
  dev.off()

  loginfo("\tPart b.")
  groups = PartB(us_arrests)
  sink("../results/R/ch10_ex9_part_b.csv")
  for (i in c(1:length(groups))) {
    print(paste0("Group ", i, ":"))
    print(paste(groups[[i]], collapse = ','))
    cat('\n')
  }
  sink()

  loginfo("\tPart c.")
  png("../results/R/ch10_ex9_part_c.png", width=FIGSIZE[1], height=FIGSIZE[2])
  PartC(us_arrests)
  dev.off()
}
